openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -469,45 +473,73 @@ def _verify_user(request, response, userid_in_session):
 
             if request_user_object_mismatch or session_user_mismatch:
                 # Log accumulated information stored on request for each change of user
-                if hasattr(request, 'debug_user_changes'):
-                    log.warning('An unsafe user transition was found. It either needs to be fixed or exempted.\n {}'
-                                .format('\n'.join(request.debug_user_changes)))
+                extra_logs = []
+
+                response_session_id = getattr(getattr(request, 'session', None), 'session_key', None)
+
+                # A safe-session user mismatch could be caused by the
+                # wrong session being retrieved from cache. This
+                # additional logging should reveal any such mismatch
+                # (without revealing the actual session ID in logs).
+                sessions_raw = [
+                    ('parsed_cookie', request.cookie_session_field),
+                    ('at_request', request.safe_cookie_verified_session_id),
+                    ('at_response', response_session_id),
+                ]
+                # Note that this is an ordered list of pairs, not a
+                # dict, so that the output order is consistent.
+                session_hashes = [(k, obscure_token(v)) for (k, v) in sessions_raw]
+                session_id_changed = len(set(kv[1] for kv in sessions_raw)) > 1
```

Comment:
Do we need to ignore None?

Reply:
I don't think so -- `obscure_token` short-circuits on `None`, if that's what you're asking about. And session changing to/from `None` is something that we should count as a change.

Reply:
For our purposes, I think we don't really care about session changes to/from `None`. Or to put it another way, I'd want a simple way to know the transactions that have a session change between two non-`None` hashes.

Reply:
Only the login/logout/registration views should normally cause a session change (and those skip this whole section), and there should be few enough of these that we're inspecting them individually anyhow.